testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs

```diff
@@ -0,0 +1,547 @@
+// Copyright (C) 2024 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::Receiver;
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::Duration;
+
+use hashbrown::{HashMap, HashSet};
+use libsigner::v0::messages::{BlockAccepted, BlockResponse, SignerMessage as SignerMessageV0};
+use libsigner::SignerEvent;
+use stacks::burnchains::Burnchain;
+use stacks::chainstate::burn::BlockSnapshot;
+use stacks::chainstate::nakamoto::NakamotoBlockHeader;
+use stacks::chainstate::stacks::boot::{NakamotoSignerEntry, RewardSet, SIGNERS_NAME};
+use stacks::chainstate::stacks::events::StackerDBChunksEvent;
+use stacks::chainstate::stacks::Error as ChainstateError;
+use stacks::types::chainstate::StacksPublicKey;
+use stacks::types::PublicKey;
+use stacks::util::hash::{MerkleHashFunc, Sha512Trunc256Sum};
+use stacks::util::secp256k1::MessageSignature;
+
+use super::Error as NakamotoNodeError;
+use crate::event_dispatcher::StackerDBChannel;
+
+/// Fault injection flag to prevent the miner from seeing enough signer signatures.
+/// Used to test that the signers will broadcast a block if it gets enough signatures
+#[cfg(test)]
+pub static TEST_IGNORE_SIGNERS: std::sync::Mutex<Option<bool>> = std::sync::Mutex::new(None);
```

Comment:
Maybe use `TestFlag` for this?

Reply:
I did not know that existed 😀.

Reply:
Refactored in de1881b5f48810a4db0b608c87036306185c18b4